File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj_list = defaultdict(list)

        for u, v in prerequisites:
            adj_list[v].append(u)
        

        in_degree = [0] * numCourses

        for prereq in range(numCourses):
            for crs in adj_list[prereq]:
                in_degree[crs] += 1
        
        queue = deque(i for i in range(numCourses) if in_degree[i] == 0)

        topo = []
        while queue:
            cur = queue.popleft()
            topo.append(cur)

            for crs in adj_list[cur]:
                in_degree[crs] -= 1

                if in_degree[crs] == 0:
                    queue.append(crs)
        

        return topo if len(topo) == numCourses else []

File: 0210-course-schedule-ii/0210-course-schedule-ii.go
func findOrder(numCourses int, prerequisites [][]int) []int {
    adj_list := make(map[int][]int, numCourses)

    for _, pair := range prerequisites {
        u, v := pair[1], pair[0]

        adj_list[u] = append(adj_list[u], v)
    }

    inDegree := make([]int, numCourses)

    for _, v := range adj_list {
        for _, crs := range v {
            inDegree[crs]++
        }
    }


    queue := make([]int, 0)

    for crs, v := range inDegree {
        if v == 0 {
            queue = append(queue, crs)
        }
    }

    topo := []int{}

    for len(queue) > 0 {
        cur := queue[0]
        queue = queue[1:]

        topo = append(topo, cur)

        for _, neigh := range adj_list[cur] {
            inDegree[neigh]--
            if inDegree[neigh] == 0 {
                queue = append(queue, neigh)
            }
        }
    }

    if len(topo) != numCourses {
        return []int{}
    }

    return topo
}